src/casp/services/api/services/consensus_engine/strategies.py

```diff
@@ -0,0 +1,230 @@
+
+    :param normalize: Boolean flag indicating whether to normalize the weights.
+    :param prune_threshold: Threshold for pruning weights below a certain value. If 0, no pruning is performed.
+    :param cell_ontology_resource: Cell ontology resource object.
+    :param cell_operations_dm: Cell operations data manager object.
+    """
+
+    REQUIRED_CELL_INFO_FEATURE_NAMES = ["cas_cell_index", "cell_type", "cell_type_ontology_term_id"]
+
+    def __init__(
+        self,
+        normalize: bool = True,
+        prune_threshold: float = 0.1,
+        cell_ontology_resource: t.Optional[CellOntologyResource] = None,
+        cell_operations_dm: t.Optional[CellOperationsDataManager] = None,
+    ):
+        self.cell_ontology_resource = cell_ontology_resource or CellOntologyResource()
+        self.cell_operations_dm = cell_operations_dm or CellOperationsDataManager()
+        self.normalize = normalize
+        self.prune_threshold = prune_threshold
+
+    def _get_ontology_aware_neighborhood_context(
+        self,
+        neighbors: t.List[MatchResult.Neighbor],
+        neighbors_metadata_dict: t.Dict[str, schemas.CellariumCellMetadata],
+    ) -> t.List[schemas.NeighborhoodSummaryOntologyAware]:
+        """
+        Utilize the ontology-aware method to assign weights to neighbor cells based on their distance and cell type
+        ontology, to inform context summarization.
+
+        :param neighbors: Neighbors of the query cell as determined by the matching engine.
+        :param neighbors_metadata_dict: Metadata for each neighbor cell.
+        :return: A list of `schemas.AnnotationInfoOntologyAware` instances, each representing the weighted cell type
+             ontology term for a neighbor cell.
+        """
+
+        neighbor_distances = np.asarray([neighbor.distance for neighbor in neighbors])
+
+        neighbor_metadata = [neighbors_metadata_dict[neighbor.cas_cell_index] for neighbor in neighbors]
+
+        # Get weights for each neighbor
+        gamma = 1.0 / np.median(neighbor_distances)
+        weights = np.exp(gamma * neighbor_distances)
+
+        total_weight = 0
+        total_usable_neighbors = 0
+        scores_dict = {k: 0 for k in self.cell_ontology_resource.ancestors_dictionary.keys()}
+
+        for neighbor_metadata, weight in zip(neighbor_metadata, weights):
+            # Cell Ontology IDs have the formats: CL:0000000 (in our database) and CL_0000000 (in the ontology graph)
+            neighbor_cell_type_ontology_id = neighbor_metadata.cell_type_ontology_term_id.replace(":", "_")
+
+            total_weight += weight
+            total_usable_neighbors += 1
+
+            # Add weight to the neighbor's ontology ID
+            scores_dict[neighbor_cell_type_ontology_id] += weight
+
+            # Propagate the weight to all ancestors (consistent subgraph of the cell type ontology)
+            ancestor_ontology_ids = self.cell_ontology_resource.ancestors_dictionary[neighbor_cell_type_ontology_id]
```

Comment:
intersting that the dictionary is working faster than array slicing...

Reply:
I think it's because getting an element from a dictionary is O(1) while slicing an array is O(n). But this was surprising for me as well, because it was a sparse array, which means that n is not number of elements, but number of non zero elements. But, anyway, I guess even with the sparse array it still affected the computation as it happens in the loop while iterating over a large number of neighbors.